### PR TITLE
[extractor/facebook] Fix Reels Format

### LINKED ISSUE #7469: [facebook:reel] ERROR No video formats found! when downloading Facebook Reels
### DO NOT REMOVE OR SKIP THE ISSUE TEMPLATE

- [X] I understand that I will be **blocked** if I *intentionally* remove or skip any mandatory\* field

### Checklist

- [X] I'm reporting that yt-dlp is broken on a **supported** site
- [X] I've verified that I'm running yt-dlp version **2023.06.22** ([update instructions](https://github.com/yt-dlp/yt-dlp#update)) or later (specify commit)
- [X] I've checked that all provided URLs are playable in a browser with the same IP and same login details
- [X] I've checked that all URLs and arguments with special characters are [properly quoted or escaped](https://github.com/yt-dlp/yt-dlp/wiki/FAQ#video-url-contains-an-ampersand--and-im-getting-some-strange-output-1-2839-or-v-is-not-recognized-as-an-internal-or-external-command)
- [X] I've searched [known issues](https://github.com/yt-dlp/yt-dlp/issues/3766) and the [bugtracker](https://github.com/yt-dlp/yt-dlp/issues?q=) for similar issues **including closed ones**. DO NOT post duplicates
- [X] I've read the [guidelines for opening an issue](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#opening-an-issue)
- [X] I've read about [sharing account credentials](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#are-you-willing-to-share-account-details-if-needed) and I'm willing to share it if required

### Region

Kenya

### Provide a description that is worded well enough to be understood

After viewing a Facebook reel on the website (the URL format is as follows:  https://www.facebook.com/reel/682865820350105) I attempted to download the video using `yt-dlp` however I got the folloing error response:

```zsh

$ ./yt-dlp "https://www.facebook.com/reel/682865820350105"
[facebook:reel] Extracting URL: https://www.facebook.com/reel/682865820350105
[facebook] Extracting URL: https://m.facebook.com/watch/?v=682865820350105&_rdr
[facebook] 682865820350105: Downloading webpage
ERROR: [facebook] 682865820350105: No video formats found!; please report this issue on  https://github.com/yt-dlp/yt-dlp/issues?q= , filling out the appropriate issue template. Confirm you are on the latest version using  yt-dlp -U

$ ./yt-dlp --version
2023.06.22

```

Also I attempted to view the reel again but in an Incognito Tab to confirm that the video is public and it worked: i.e. I could view the video, meaning it is a public URL not needed Facebook credentials to view.

NOTE: Using the command line utility, and NOT using the API.

### Provide verbose output that clearly demonstrates the problem

- [X] Run **your** yt-dlp command with **-vU** flag added (`yt-dlp -vU <your command line>`)
- [X] If using API, add `'verbose': True` to `YoutubeDL` params instead
- [X] Copy the WHOLE output (starting with `[debug] Command-line config`) and insert it below

### Complete Verbose Output

```shell
[debug] Command-line config: ['-vU', 'https://www.facebook.com/reel/682865820350105']
[debug] Encodings: locale UTF-8, fs utf-8, pref UTF-8, out utf-8, error utf-8, screen utf-8
[debug] yt-dlp version stable@2023.06.22 [812cdfa06] (zip)
[debug] Python 3.10.7 (CPython x86_64 64bit) - Linux-5.19.0-43-generic-x86_64-with-glibc2.36 (OpenSSL 3.0.5 5 Jul 2022, glibc 2.36)
[debug] exe versions: ffmpeg 5.1.1 (setts), ffprobe 5.1.1
[debug] Optional libraries: certifi-2022.12.07, secretstorage-3.3.3, sqlite3-2.6.0
[debug] Proxy map: {}
[debug] Loaded 1851 extractors
[debug] Fetching release info: https://api.github.com/repos/yt-dlp/yt-dlp/releases/latest
Available version: stable@2023.06.22, Current version: stable@2023.06.22
Current Build Hash: 924464eb831bdb99d854b9d47ff9e1c0870eab276113adaf8a74a8561ad332b9
yt-dlp is up to date (stable@2023.06.22)
[facebook:reel] Extracting URL: https://www.facebook.com/reel/682865820350105
[facebook] Extracting URL: https://m.facebook.com/watch/?v=682865820350105&_rdr
[facebook] 682865820350105: Downloading webpage
ERROR: [facebook] 682865820350105: No video formats found!; please report this issue on  https://github.com/yt-dlp/yt-dlp/issues?q= , filling out the a
ppropriate issue template. Confirm you are on the latest version using  yt-dlp -U
Traceback (most recent call last):
  File "/home/emk/Developer/SCRATCH/youtube-downloader/./yt-dlp/yt_dlp/YoutubeDL.py", line 1555, in wrapper
    return func(self, *args, **kwargs)
  File "/home/emk/Developer/SCRATCH/youtube-downloader/./yt-dlp/yt_dlp/YoutubeDL.py", line 1652, in __extract_info
    return self.process_ie_result(ie_result, download, extra_info)
  File "/home/emk/Developer/SCRATCH/youtube-downloader/./yt-dlp/yt_dlp/YoutubeDL.py", line 1711, in process_ie_result
    ie_result = self.process_video_result(ie_result, download=download)
  File "/home/emk/Developer/SCRATCH/youtube-downloader/./yt-dlp/yt_dlp/YoutubeDL.py", line 2653, in process_video_result
    self.raise_no_formats(info_dict)
  File "/home/emk/Developer/SCRATCH/youtube-downloader/./yt-dlp/yt_dlp/YoutubeDL.py", line 1067, in raise_no_formats
    raise ExtractorError(msg, video_id=info['id'], ie=info['extractor'],
yt_dlp.utils.ExtractorError: [facebook] 682865820350105: No video formats found!; please report this issue on  https://github.com/yt-dlp/yt-dlp/issues?
q= , filling out the appropriate issue template. Confirm you are on the latest version using  yt-dlp -U
```

### PR DESCRIPTION
**Pull request by demon071**
**IMPORTANT**: PRs without the template will be CLOSED

### Description of your *pull request* and other information

Get video format for facebook reels

Fixes #

<details open><summary>Template</summary> 

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8) and [ran relevant tests](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check all of the following options that apply:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [x] Fix or improvement to an extractor (Make sure to add/update tests)
- [ ] New extractor ([Piracy websites will not be accepted](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#is-the-website-primarily-used-for-piracy))
- [ ] Core bug fix/improvement
- [ ] New feature (It is strongly [recommended to open an issue first](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#adding-new-feature-or-making-overarching-changes))

</details><details><summary>Copilot Summary</summary>  

### Summary
🎥🛠️🔗

Improve Facebook extractor by adding a fallback for playable_url. Use the `creation_story` field of the video object to get the `playable_url` if the original field is missing. This fixes some videos not being downloadable.

> _Some videos fail_
> _`playable_url` missing_
> _Fall back to `story`_

### Walkthrough
* Add support for more video types, such as live videos, premiere videos, and videos from groups or pages ([link](https://github.com/yt-dlp/yt-dlp/pull/7564/files?diff=unified&w=0#diff-ca6052d2c84eb4ac45aaf04463a4eaecdc75fb3ae4766e234917d8ab131b60ecL504-R504),                            F0

</details>

### CODE REVIEW

File: yt_dlp/extractor/facebook.py [hunk 1]
```diff
@@ -501,7 +501,7 @@ def parse_graphql_video(video):
                     q = qualities(['sd', 'hd'])
                     for key, format_id in (('playable_url', 'sd'), ('playable_url_quality_hd', 'hd'),
                                            ('playable_url_dash', '')):
-                        playable_url = video.get(key)
+                        playable_url = video.get(key) or video.get('creation_story', {}).get('short_form_video_context', {}).get('playback_video', {}).get(key)
```
Comment (pukkandan) [inline yt_dlp/extractor/facebook.py:504]:
```suggestion
                        playable_url = traverse_obj(
                            video, key, ('creation_story', 'short_form_video_context', 'playback_video', key))
```

**Review by pukkandan — CHANGES_REQUESTED** [review]
Add or update tests

**Comment by pukkandan** [comment]
Just push the md5 change and let the metadata tests fail. We can merge anyway

**Comment by bashonly** [comment]
Would it be ok to comment out the missing fields with a `TODO` so that it's known what needs to be fixed?

**Comment by pukkandan** [comment]
Yeah, don't comment it. Tests that pretend to succeed is worse than failing tests.

**Comment by bashonly** [comment]
we could do this to get the metadata. it's ugly, but this extractor has kinda painted itself into a corner, and i'm not seeing a better way to do it without a significant rewrite
```diff
diff --git a/yt_dlp/extractor/facebook.py b/yt_dlp/extractor/facebook.py
index 9d871eb28..40c144c0b 100644
--- a/yt_dlp/extractor/facebook.py
+++ b/yt_dlp/extractor/facebook.py
@@ -497,6 +498,13 @@ def extract_relay_prefetched_data(_filter):
                 entries = []
 
                 def parse_graphql_video(video):
+                    v_id = video.get('videoId') or video.get('id') or video_id
+                    reel_info = traverse_obj(
+                        video, ('creation_story', 'short_form_video_context', 'playback_video', {dict}))
+                    if reel_info:
+                        video = video['creation_story']
+                        video['owner'] = traverse_obj(video, ('short_form_video_context', 'video_owner'))
+                        video.update(reel_info)
                     formats = []
                     q = qualities(['sd', 'hd'])
                     for key, format_id in (('playable_url', 'sd'), ('playable_url_quality_hd', 'hd'),
@@ -513,15 +521,15 @@ def parse_graphql_video(video):
                                 'url': playable_url,
                             })
                     extract_dash_manifest(video, formats)
-                    v_id = video.get('videoId') or video.get('id') or video_id
                     info = {
                         'id': v_id,
                         'formats': formats,
                         'thumbnail': traverse_obj(
                             video, ('thumbnailImage', 'uri'), ('preferred_thumbnail', 'image', 'uri')),
                         'uploader_id': try_get(video, lambda x: x['owner']['id']),
-                        'timestamp': int_or_none(video.get('publish_time')),
-                        'duration': float_or_none(video.get('playable_duration_in_ms'), 1000),
+                        'timestamp': traverse_obj(video, 'publish_time', 'creation_time', expected_type=int_or_none),
+                        'duration': (float_or_none(video.get('playable_duration_in_ms'), 1000)
+                            or float_or_none(video.get('length_in_second'))),
                     }
                     process_formats(info)
                     description = try_get(video, lambda x: x['savable_description']['text'])
```

**Review by pukkandan — APPROVED** [review]

**Comment by briankung** [comment]
Just tested this branch on a Reels video and it worked. Thanks for the hard work!

**Pull request merged**